File: ui/components/ui/ProducerSession/message-generator/data-generators/json-generator/json-generator.ts
import * as pb from '../../../../../../grpc-web/tools/teal/pulsar/ui/producer/v1/producer_pb';
import { FixedJsonGenerator, fixedJsonGeneratorFromPb, fixedJsonGeneratorToPb } from './fixed-json-generator';
import { JsJsonGenerator, jsJsonGeneratorFromPb, jsJsonGeneratorToPb } from './js-json-generator';

export type JsonGenerator = {
  generator: FixedJsonGenerator | JsJsonGenerator
}

export function jsonGeneratorFromPb(v: pb.JsonGenerator): JsonGenerator {
  let generator: JsonGenerator['generator'];
  switch (v.getGeneratorCase()) {
    case pb.JsonGenerator.GeneratorCase.GENERATOR_FIXED:
      generator = fixedJsonGeneratorFromPb(v.getGeneratorFixed()!);
      break;
    case pb.JsonGenerator.GeneratorCase.GENERATOR_JS:
      generator = jsJsonGeneratorFromPb(v.getGeneratorJs()!);
      break;
    default:
      throw new Error(`Unknown json generator: ${v.getGeneratorCase()}`);
  }

  return { generator };
}

export function jsonGeneratorToPb(v: JsonGenerator): pb.JsonGenerator {
  const message = new pb.JsonGenerator();
  if ('json' in v.generator) {
    message.setGeneratorFixed(fixedJsonGeneratorToPb(v.generator));
  } else {
    message.setGeneratorJs(jsJsonGeneratorToPb(v.generator));
  }

  return message;
}

File: server/src/main/scala/producer/message_generator/data_generators/json_generator/JsonGenerator.scala
package producer.message_generator.data_generators.json_generator

import com.tools.teal.pulsar.ui.producer.v1.producer as pb
import org.graalvm.polyglot.Context

case class JsonGenerator(
    generator: FixedJsonGenerator | JsJsonGenerator
):
    def generate(polyglotContext: Context): String =
        generator match
            case v: FixedJsonGenerator => v.generate
            case v: JsJsonGenerator => v.generate(polyglotContext)

object JsonGenerator:
    def fromPb(v: pb.JsonGenerator): JsonGenerator =
        val generator  = v.generator match
            case pb.JsonGenerator.Generator.GeneratorFixed(v) => FixedJsonGenerator.fromPb(v)
            case pb.JsonGenerator.Generator.GeneratorJs(v) => JsJsonGenerator.fromPb(v)
            case _ => throw new Exception("Unknown generator type")

        JsonGenerator(
            generator = generator
        )

    def toPb(v: JsonGenerator): pb.JsonGenerator =
        val generator = v.generator match
            case v: FixedJsonGenerator => pb.JsonGenerator.Generator.GeneratorFixed(FixedJsonGenerator.toPb(v))
            case v: JsJsonGenerator => pb.JsonGenerator.Generator.GeneratorJs(JsJsonGenerator.toPb(v))
        pb.JsonGenerator(
            generator = generator
        )
